### PR TITLE
Various Improvements

### LINKED ISSUE #18: generate_features_for_folder not found in repo
The function used in the example in README.md doesnt actually exist .. 
Feel free to close this issue if it's part of the "complete notebook" other issue..
Thanks!

### PR DESCRIPTION
**Pull request by kahrendt**
A large set of changes here! I still need to clean up a few things and verify it is all working in a scratch environment before merging.

- Audio augmentation and spectrogram generation has been moved into several different classes
- Spectrogram features are stored in uint16 format to save space. It should automatically be converted to float features while training or inferring.
- Generated samples can be augmented 'on-the-fly' while training (this allows easy experimentation of various settings)
- Removed FreqMix and MixUp augmentations, as they did not improve models in my testing.
- Data loading while training has been refactored to better support live augmentations and to simplify logic. These changes will potentially support generating audio samples 'on-the-fly' as well.
- Cleaned up the code that converts the non-streaming models to streaming models
- Added a StridedKeep layer. This allows us to use one streaming ring buffer when the data splits into different branches in the model. This reduces overhead and decreases the amount of memory needed for the tensor arena when running on TFLM.
- Added a new model architecture called MixedNet. This is much faster than the Inception models while using less parameters (it is even fast enough to run on plain ESP32 devices). Hyperparameters can be better optimized!
- Added a new metric to help choose the best model from training. It computes the average recall for false accept per hour rates between 0 and 2 minutes. The code is still hacked together and can be improved.
- Add helper functions for generating ROC curves as well as computing false accept per hour rates at various thresholds. ~~They are currently unused, but I will eventually use them to recommend various probability cutoff thresholds after a model is trained.~~
- Most functions are properly typed and comments have been improved.
- Increased the version in ``setup.py`` and added the webrtcvad package. Currently microWakeWord is targeting TensorFlow 2.14 and 2.15, as 2.16 brings many changes to Keras.
- The notebooks have been updated to show the new spectrogram generation and training process. The feature generation notebook will also downloaded some precomputed spectrogram features I have uploaded to Hugging Face. The missing piece for an end-to-end training notebook is sample generation.
- Readme has been updated to reflect many of these changes (will close #18 when merged)
- When testing a trained model, it outputs the false rejection and false accept per hour rates at various probability cutoffs to help determine the best setting.

Here is what I still need to do before merging:
- Setup the environment from scratch and train a new model to make sure all the notebooks are working as expected.
- ~~Use the new helper functions to recommend various probability thresholds after a model is done training.~~
- The new average viable recall code should be cleaned up.
- Fix any other issues/errors that have popped up in the process.

### CODE REVIEW

**Comment by muik** [comment]
@kahrendt I'm really looking forward to the successful merge of these improvements. Thank you for all your hard work!
